blockchain/state_migration.go

```diff
@@ -528,3 +531,136 @@ func (bc *BlockChain) StopWarmUp() error {
+func (bc *BlockChain) StartCollectingTrieStats(contractAddr common.Address) error {
+	block := bc.GetBlockByNumber(bc.lastCommittedBlock)
+	if block == nil {
+		return fmt.Errorf("Block #%d not found", bc.lastCommittedBlock)
+	}
+
+	mainTrieDB := bc.StateCache().TrieDB()
+	cache := mainTrieDB.TrieNodeCache()
+	if cache == nil {
+		return fmt.Errorf("target cache is nil")
+	}
+	db := state.NewDatabaseWithExistingCache(bc.db, cache)
+
+	startNode := block.Root()
+	// If the contractAddr is given, start collecting stats from the root of storage trie
+	if !common.EmptyAddress(contractAddr) {
+		var err error
+		startNode, err = bc.GetContractStorageRoot(block, db, contractAddr)
+		if err != nil {
+			logger.Error("Failed to get the contract storage root",
+				"contractAddr", contractAddr.String(), "rootHash", block.Root().String(),
+				"err", err)
+			return err
+		}
+	}
+
+	children, err := db.TrieDB().NodeChildren(startNode)
+	if err != nil {
+		logger.Error("Failed to retrieve the children of start node", "err", err)
+		return err
+	}
+
+	logger.Info("Started collecting trie statistics",
+		"blockNum", block.NumberU64(), "root", block.Root().String(), "len(children)", len(children))
+	go collectTrieStats(db, startNode)
+
+	return nil
+}
+
+// collectChildrenStats wraps CollectChildrenStats, in order to send finish signal to resultCh.
+func collectChildrenStats(db state.Database, child common.Hash, resultCh chan<- statedb.NodeInfo) {
+	db.TrieDB().CollectChildrenStats(child, 2, resultCh)
+	resultCh <- statedb.NodeInfo{Finished: true}
+}
+
+// collectTrieStats is the main function of collecting trie statistics.
+// It spawns goroutines for the upper-most children and iterates each sub-trie.
+func collectTrieStats(db state.Database, startNode common.Hash) {
+	children, err := db.TrieDB().NodeChildren(startNode)
+	if err != nil {
+		logger.Error("Failed to retrieve the children of start node", "err", err)
+	}
+
+	// collecting statistics by running individual goroutines for each child
+	resultCh := make(chan statedb.NodeInfo, 10000)
+	for _, child := range children {
+		go collectChildrenStats(db, child, resultCh)
+	}
+
+	numGoRoutines := len(children)
+	ticker := time.NewTicker(1 * time.Minute)
+
+	numNodes, numLeafNodes, maxDepth := 0, 0, 0
+	depthCounter := make(map[int]int)
+	begin := time.Now()
+	for {
+		select {
+		case result := <-resultCh:
+			if result.Finished {
+				numGoRoutines--
+				if numGoRoutines == 0 {
+					logger.Info("Finished collecting trie statistics", "elapsed", time.Since(begin),
+						"numNodes", numNodes, "numLeafNodes", numLeafNodes, "maxDepth", maxDepth)
+					printDepthStats(depthCounter)
+					return
+				}
+				continue
+			}
+			numNodes++
+			// if a leaf node, collect the depth data
+			if result.Depth != 0 {
+				numLeafNodes++
+				depthCounter[result.Depth]++
+				if result.Depth > maxDepth {
+					maxDepth = result.Depth
+				}
+			}
+		case <-ticker.C:
```

Comment:
Can `case <-bc.quit` stop this loop?
If this API can take too long, we need to return this function by `case <-bc.quit`.

Reply:
Doesn't `bc.quit` mean turn down the process of a node? If we need to gracefully kill this operation, that might be needed. However, as it is a read-only operation, I think that won't be necessarily needed. How do you think? If we're worried about the long-running operation, we'd better have an another API to send signal to this operation.

Reply:
Generally, I think we need to consider graceful kill operation. But this is debug API. It is not very important to consider it. It's ok to go.

Sorry for the delay in review.